pkg/sql/schemachanger/scbuild/builder_state.go

```diff
@@ -230,6 +230,30 @@ func (b *builderState) NextViewIndexID(view *scpb.View) (ret catid.IndexID) {
 	return b.nextIndexID(view.ViewID)
 }
 
+func (b *builderState) NextConstraintID(table *scpb.Table) (ret catid.ConstraintID) {
+	{
+		b.ensureDescriptor(table.TableID)
+		desc := b.descCache[table.TableID].desc
+		tbl, ok := desc.(catalog.TableDescriptor)
+		if !ok {
+			panic(errors.AssertionFailedf("Expected table descriptor for ID %d, instead got %s",
+				desc.GetID(), desc.DescriptorType()))
+		}
+		ret = tbl.GetNextConstraintID()
+	}
+	// We assume that every constraint comes with a ConstraintName element,
+	// so it suffices to iterate all constraint names to figure out the max
+	// constraint ID used so far.
+	scpb.ForEachConstraintName(b.QueryByID(table.TableID), func(
+		_ scpb.Status, _ scpb.TargetStatus, constraintName *scpb.ConstraintName,
+	) {
+		if constraintName.ConstraintID >= ret {
+			ret = constraintName.ConstraintID + 1
+		}
+	})
```

Comment:
Drive-by reviewing: you should be able to walk through all the constraint IDs in the table by doing a little rel-magic:
```
	nextID := catid.ConstraintID(1)
	b.QueryByID(table.TableID).ForEachElementStatus(func(_ scpb.Status, _ scpb.TargetStatus, e scpb.Element) {
		v, _ := screl.Schema.GetAttribute(screl.ConstraintID, e)
		if id, ok := v.(catid.ConstraintID); ok && id >= nextID {
			nextID = id + 1
		}		
	})
```
This should allow you to remove this method entirely, which would be nice.

Reply:
Thanks! We have similar nextIndexID, nextColumnID method and I like having a nextConstraintID method as well (since it might be used more than once). But I do like and will change the implementation to what you suggested.

Reply:
edit: in fact this method is never used in this PR (but will be in the next followup PR) so I delete it for now.